lib/valkyrie/storage/memory.rb

```diff
@@ -30,11 +33,54 @@ def handles?(id:)
       id.to_s.start_with?("memory://")
     end
 
-    # Delete the file on disk associated with the given identifier.
+    # Delete the file in memory associated with the given identifier.
     # @param id [Valkyrie::ID]
     def delete(id:)
       cache.delete(id)
       nil
     end
+
+    def supports_versions?
+      true
+    end
+
+    # @param id [Valkyrie::ID]
+    def versions(id:)
+      uri = URI(id.to_s)
+      # This already is a version, so just return itself.
+      return [id] if uri.path.present?
+      resource_id = uri.host
+      cache[to_id(resource_id: resource_id)].keys.map do |v|
+        Valkyrie::ID.new("memory://#{resource_id}/#{v}")
+      end
+    end
+
+    private
+
+      def put(key, value)
+        id = to_id(key)
+        cache[id] ||= {}
+        cache[id][key[:version]] = value
+      end
+
+      def head_version(id:)
+        object = cache[id]
+        return unless object
+        object.keys.max
+      end
+
+      def next_version_key(resource:)
+        id = to_id(resource_id: resource.id)
+        n = if cache.key?(id)
+              cache[id].keys.count
+            else
+              0
+            end
+        { resource_id: resource.id, version: n }
+      end
+
+      def to_id(key)
```

Comment:
Given the above method `to_valkyrie_id` consider renaming this method `to_valkyrie_id` for a more consistent expectation.